File: AtCoder/AGC021/D.rs
use std::io::Read;

fn main() {
    let mut buf = String::new();
    std::io::stdin().read_to_string(&mut buf).ok();
    let mut it = buf.split_whitespace();
    let s = it.next().unwrap().as_bytes();
    let n = s.len();
    let k = it.next().unwrap().parse::<usize>().unwrap();
    let mut dp = vec![vec![vec![0; n]; n]; k+1];
    for d in (1..n).rev() {
        for i in 0..k+1 {
            for j in 0..n-d {
                dp[i][j+1][j+d] = std::cmp::max(dp[i][j+1][j+d], dp[i][j][j+d]);
                dp[i][j][j+d-1] = std::cmp::max(dp[i][j][j+d-1], dp[i][j][j+d]);
                let next_i = i + if s[j] == s[j+d] { 0 } else { 1 };
                if next_i > k { continue; }
                dp[next_i][j+1][j+d-1] = std::cmp::max(dp[next_i][j+1][j+d-1], dp[i][j][j+d] + 2);
            }
        }
    }
    let mut res = 0;
    for i in 0..k+1 {
        for j in 0..n {
            for l in 0..n {
                res = std::cmp::max(res, dp[i][j][l] + if j == l { 1 } else { 0 });
            }
        }
    }
    println!("{}", res);
}

File: AtCoder/AGC021/D.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int main(){
    string S; cin >> S;
    const int N = S.size();
    int K; cin >> K;
    int res = 0;
    vector<vector<vector<int>>> dp(K+1, vector<vector<int>>(N, vector<int>(N, 0)));
    for(int d=N-1;d>=0;d--){
        for(int i=0;i+d<N;i++){
            int j = i+d;
            for(int k=0;k<=K;k++){
                if(k < K && S[i] != S[j]){
                    res = max(res, dp[k][i][j] + 2);
                    if(i+1 < N && j > 0){
                        dp[k+1][i+1][j-1] = max(dp[k+1][i+1][j-1], dp[k][i][j] + 2);
                    }
                }
                if(S[i] == S[j]){
                    res = max(res, dp[k][i][j] + (i==j ? 1 : 2));
                    if(i+1 < N && j > 0){
                        dp[k][i+1][j-1] = max(dp[k][i+1][j-1], dp[k][i][j] + (i==j ? 1 : 2));
                    }
                }
                if(i+1 < N) dp[k][i+1][j] = max(dp[k][i+1][j], dp[k][i][j]);
                if(j > 0) dp[k][i][j-1] = max(dp[k][i][j-1], dp[k][i][j]);
            }
        }
    }
    cout << res << endl;
}